### ISSUE TITLE
Linut Mint /etc/os-release bug causes failures in Advantage APT and MOTD Messages

### ISSUE THREAD

**Issue by vossenjp** [linuxmint/linuxmint#392]
Several weeks ago I started getting these failures in my logs:
```
Jun 18 03:16:22 hostname python3[3568780]: Traceback (most recent call last):
Jun 18 03:16:22 hostname python3[3568780]:   File "/usr/lib/ubuntu-advantage/ua_update_messaging.py", line 302, in <module>
Jun 18 03:16:22 hostname python3[3568780]:     update_apt_and_motd_messages(cfg=cfg)
Jun 18 03:16:22 hostname python3[3568780]:   File "/usr/lib/ubuntu-advantage/ua_update_messaging.py", line 283, in update_apt_and_motd_messages
Jun 18 03:16:22 hostname python3[3568780]:     series = util.get_platform_info()["series"]
Jun 18 03:16:22 hostname python3[3568780]:   File "/usr/lib/python3/dist-packages/uaclient/util.py", line 331, in get_platform_info
Jun 18 03:16:22 hostname python3[3568780]:     raise RuntimeError(
Jun 18 03:16:22 hostname python3[3568780]: RuntimeError: Could not parse /etc/os-release VERSION: 20 (Ulyana) (modified to 20 (Ulyana))
Jun 18 03:16:22 hostname systemd[1]: ua-messaging.service: Main process exited, code=exited, status=1/FAILURE
Jun 18 03:16:22 hostname systemd[1]: ua-messaging.service: Failed with result 'exit-code'.
Jun 18 03:16:22 hostname systemd[1]: Failed to start Ubuntu Advantage APT and MOTD Messages.
```

This is caused by:
```
$ cat /etc/os-release 
NAME="Linux Mint"
VERSION="20 (Ulyana)"                  # <<<<<<< THIS LINE should be 20.0
ID=linuxmint
ID_LIKE=ubuntu
PRETTY_NAME="Linux Mint 20"
VERSION_ID="20"
HOME_URL="https://www.linuxmint.com/"
SUPPORT_URL="https://forums.linuxmint.com/"
BUG_REPORT_URL="http://linuxmint-troubleshooting-guide.readthedocs.io/en/latest/"
PRIVACY_POLICY_URL="https://www.linuxmint.com/"
VERSION_CODENAME=ulyana
UBUNTU_CODENAME=focal
```

Because `/usr/lib/python3/dist-packages/uaclient/util.py` (from "ubuntu-advantage-tools") has:
```
...
     44 REGEX_OS_RELEASE_VERSION = r"(?P<release>\d+\.\d+) (LTS )?\((?P<series>\w+).*"
...
    329     match = re.match(REGEX_OS_RELEASE_VERSION, version)
    330     if not match:
    331         raise RuntimeError(
    332             "Could not parse /etc/os-release VERSION: {} (modified to"
    333             " {})".format(os_release["VERSION"], version)
    334         )
...
```
And clearly `VERSION="20 (Ulyana)"` does not contain `\.\d` so it will never match `(?P<release>\d+\.\d+)`.

Personally, I've never liked the inconsistency of Mint "20" vs. "21.1", etc.  I've always thought it should start "20.0".

I've filed this here because:
```
$ apt policy base-files
base-files:
  Installed: 20.0.1
  Candidate: 20.0.1
  Version table:
 *** 20.0.1 700
        700 http://mirror.cs.jmu.edu/pub/linuxmint/packages ulyana/upstream amd64 Packages
        100 /var/lib/dpkg/status
     11ubuntu5.3 500
        500 http://mirror.math.princeton.edu/pub/ubuntu focal-updates/main amd64 Packages
     11ubuntu5 500
        500 http://mirror.math.princeton.edu/pub/ubuntu focal/main amd64 Packages
```

And I think the bug is in Mint's `os-release::VERSION` and not the regexp in "ubuntu-advantage-tools," though that could be hacked to something like `(?P<release>\d+(\.\d+)?)` also.

Needless to say, when I get around to upgrading to 20.x the problem will go away.  Until "21"...

I've just tried a short-term fix of manually editing `/etc/os-release` and changing to "20.0", but I don't know if that will break anything else yet.

**Comment by xenopeek** [comment]
I think the issue here is not that version 20 should be written as 20.0 but that the Ubuntu Advantage (UA) scripts expect to read a Ubuntu version number in os-release VERSION string and whether Linux Mint 20 is written as 20 or 20.0 this is not an Ubuntu version number.

It happens to be the case for this release that the major version of Linux Mint 20.x and its package base Ubuntu 20.4 LTS match but that's not the case for other releases. Linux Mint 19.x were based on Ubuntu 18.04 LTS (but Bionic used different UA scripts) and next year's Linux Mint 21 will be based on Ubuntu 22.4 LTS. UA will expect to read 22.4 LTS as version for that but whether Linux Mint 21 is written as 21 or 21.0 neither will give the UA scripts the Ubuntu version it expects to find.

In short replacing 20 with 20.0 would solve this regex issue but it won't solve the underlying that Ubuntu Advantage expects to be run on Ubuntu. I don't see a quick fix for this.

**Comment by vossenjp** [comment]
Well, that's a good point.  I was _assuming_ that running "Ubuntu Advantage" on Mint was on purpose, and that there were Mint overrides as needed.  I'm also hazy on the relationship between UA and "fwupd & fwupdmgr Mint Firmware Update Manager" which also seems to update MotD.

It's only been a day, but my hack of adding ".0" fixed the exception and now I get:
```
Jun 19 03:29:16 hostname systemd[1]: Finished Ubuntu Advantage APT and MOTD Messages.
```

**Comment by xenopeek** [comment]
I didn't realize ubuntu-advantage-tools was default install on the ISO. I think maybe it should be removed or, as you say, it might need some Mint override to work correctly (like checking the Ubuntu version from DISTRIB_RELEASE in `/etc/upstream-release/lsb-release` instead of from VERSION in `/etc/os-release`). The .0 hack works now but I didn't follow the entire code back from where the script failed its regex parsing of VERSION, what else is does with the major.minor version numbers. It looked to do something with it later.

**Comment by vossenjp** [comment]
Interesting!  I just ran updates and these are in there:
```
ubuntu-advantage-tools	27.0.2~18.04.1 -> 27.1~18.04.1
ubuntu-advantage-tools	27.0.2~20.04.1 -> 27.1~20.04.1
```

But no regexp change:
```
$ grep 'REGEX_OS_RELEASE_VERSION' /usr/lib/python3/dist-packages/uaclient/util.py
REGEX_OS_RELEASE_VERSION = r"(?P<release>\d+\.\d+) (LTS )?\((?P<series>\w+).*"
    match = re.match(REGEX_OS_RELEASE_VERSION, version)
```

**Comment by vossenjp** [comment]
I should note that the exceptions in my logs started a few weeks ago, only on Mint-20, but not on Mint-19.3, which makes sense given that regexp.  I did not change change my log monitoring, so I assume something changed in UA.  But when I look in https://git.launchpad.net/ubuntu/+source/ubuntu-advantage-tools/ it doesn't seem like that line has changed since:
```
21daa1b5        (Chad Smith     2020-03-30 14:49:17 -0600       45)REGEX_OS_RELEASE_VERSION = r"(?P<release>\d+\.\d+) (LTS )?\((?P<series>\w+).*"
```
Not sure where Mint's code might be, or if/when the Mint-20(.0) code might have changed.

I can't pin it down better than "a few weeks ago" because my logs have rotated off and I hadn't had time to dive into this until now, so I've lost track of when it started.  Hmmm, OK, `/var/log/apt/history.log*` have:
* Start-Date: 2021-06-19  14:16:22 = ...`ubuntu-advantage-tools:amd64 (27.0.2~20.04.1, 27.1~20.04.1)`...
* Start-Date: 2021-05-15  14:09:39 = ...`ubuntu-advantage-tools:amd64 (20.3, 27.0.2~20.04.1)`...

2021-05-15 seems about right and if I'm reading that right 20.3 to 27.0 is a pretty big jump and might be suspicious.
